### PR TITLE
Keymap Minivan: configurable dual layout, many symbols, speed/text-size measuring

### PR DESCRIPTION
**Pull request by josjoha**
## Keymap for Minivan

This is a keymap for the Minivan keyboard. It works now for all hardware variants, 12x12x12x[8, 9, 10], and might be easily ported to the Planck (12×12×12×12, untested).

It supports a dual layout, with various options for those layouts (Qwerty, Dvorak, Colemak, Numpad (done), shortly: Qwerty with arrows on top, Workman, Azerty and Qwertz).

Keys (more than standard keyboard, but not modern keys like e-mail, www). All alphanumerical keys, the usual modifiers, F-keys, navigation and mouse, power- media-keys, numpad keys, all accented characters used in western Europe, and two layers of additional Unicode symbols.

The readme.md file lists all features.

I think it's a pretty cool keymap (if I may say so), so maybe someone else likes to use it too on another keyboard. I hope this will be a good 40% keymap (intuitive, fast, capable), for people without the time to program their own.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Dual keymap for Minivan. 
* A fair amount of the more well known layouts to choose between, such as Qwerty, Dvorak, Colemak, …
* Additional symbols (Unicode).
* Typing speed measuring.
* Text size measuring.
* Hopefully easy to port to other mini keyboards, especially the Planck.
* Configurable.
* Not needing the dead-key system for accented characters (Dutch), so that vi(1) registers can be normally used.
* It became the goal to make a solid 40% keymap, for everyone without the time to do it themselves, fit for use with all programs and operating systems, for most/all of the usual user preferences, without quirks, predictable, easy to adapt to, for fast typing with 10 fingers, exhaustively documented, maintainable source code.

## Checklist

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

I do not use Qwerty at all and can therefore not test that well. I have attempted to incorporate the standard Qwerty keymap, just as was done with Dvorak.
(I'll try to be on discord QMK #programming as 'josjoha'.)

### CODE REVIEW

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/config.h
Comment (drashna) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/config.h]: Should be wrapped with `#ifndef LINK_TIME_OPTIMIZATION_ENABLE`.

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/config.h
Comment (drashna) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/config.h]: >// Did not decrease firmware size when tested on 26 Jan 2020 

That's because this should be in the rules.mk, as `LTO_ENABLE = yes`, or `LINK_TIME_OPTIMIZATION_ENABLE = yes`.  This will enable LTO for the board, and will automatically define `NO_ACTION_MACRO` and `NO_ACTION_FUNCTION`. 

https://github.com/qmk/qmk_firmware/blob/5b91c3e0a0dc8152f69130cf047f3df0d0f94421/tmk_core/common.mk#L170-L179

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c
Comment (drashna) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c]: This needs to be defined in the config.h, or it will not be picked up correctly.

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c
Comment (drashna) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c]: This isn't needed, as it's included globally. 
```suggestion
```

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c
Comment (drashna) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c]: You may want/need to move this to the config.h file, so it's applied to all files, properly.

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c
Comment (drashna) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c]: This is completely redundant.  `layer_move` does this, by turning off all layers, and then turning on just the specified layer.

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c
Comment (drashna) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c]:
```suggestion
layer_state_t layer_state_set_user (layer_state_t state) {
```

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c
Comment (josjoha) [reply keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c]: Ok, I left #include QMK_KEYBOARD_H in keymap.c because I got errors moving it to config.h, and moved #define RGBLIGHT_SLEEP to config.h, hoping I understood you correctly.

**Comment by josjoha** [comment]
Your request (thanks for that) was implemented correctly, I hope.

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c
Comment (drashna) [reply keyboards/thevankeyboards/minivan/keymaps/josjoha/keymap.c]: Yup, that's correct

**Comment by josjoha** [comment]
I'll change it to markdown tomorrow. I had it in ODT to make it easy to print. I could add one for Qwerty+Dvorak too. I could also just delete it.

**Comment by josjoha** [comment]
Been thinking a bit on adding a Minivan 'South paw' (one more key first row left) and 'Arrow' (one more key first row rigth) layout compile option. 'South paw' is easy, it just adds a key. 'Arrow' is harder, if it should function as arrows. If not that is also easy.

It seems the best way to do it with arrows is to have the additional key be a layer move key on the BASE layer to the navigation layer (_MOV). On _MOV add an additional arrow cluster in that spot. That seems to be better than to loose three keys on the BASE layer: GUI, the standard 'z' (Dvorak) or ';' in Qwerty (less bad), and the _RAR layer key. To push an arrow cluster on the BASE layer would require additional shuffling around. I don't think I would want to get into that, it be messy. 

It be helpful if someone had such a configuration to test it on, otherwise I would need to reconfigure my board; that be a bit tedious for testing back and forth. Just let me know if you think it is a good idea. The user would only have to uncomment a #define in keymap.c.

It would merely require to add more keycode definition(s) on the first layer, I assume.

**Comment by josjoha** [comment]
If u guys can hold on for a few days with merging, I will integrate an option for an arrow cluster for the 'Arrow' hardware layout, and test these layouts (44,45,46 keys) on my board. That way we avoid the overhead of having another pull request, or starting a new one.

**Comment by josjoha** [comment]
@drashna Looks like that's it. For my part it can be merged.

**Comment by josjoha** [comment]
The check went correctly, but now seems botched on someone else's code, while I only changed something in my documentation. I assume it is not my problem ?

**Comment by drashna** [comment]
It looks like you may need to run `make git-submodule` and add a new commit to fix the lib/ errors.

Otherwise, @noroadsleft ?

**Comment by josjoha** [comment]
I merged the latest, did  **make git-submodule** which did solve some messages regarding libs. It seems it still hangs on something else. No problems compiling only this keymap here.

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/config.h
Comment (tzarc) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/config.h:1]: Lets add a license header.

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/rules.mk
Comment (tzarc) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/rules.mk:14]:
```suggestion
```

File: keyboards/thevankeyboards/minivan/keymaps/josjoha/unicode_macros.h
Comment (tzarc) [inline keyboards/thevankeyboards/minivan/keymaps/josjoha/unicode_macros.h:23]:
```suggestion
#pragma once

```

Comment (josjoha) [reply keyboards/thevankeyboards/minivan/keymaps/josjoha/unicode_macros.h:23]: Thanks for reviewing, I'll incorporate it once I can. Still doing some minor things, then I'll try an upload.
I wouldn't recommend digging through the keymap at this time, because so much has changed.

**Comment by josjoha** [comment]
That should be it, notwithstanding some more alpha testing. The firmware size was about 350-410 bytes left over, before I pulled in the latest QMK. With the latest QMK this dropped to 150 bytes free. (Debian avr-gcc 5.4.0)
I don't know what to do about errors with other keyboards.

**Comment by drashna** [comment]
> I don't know what to do about errors with other keyboards.

The cause of that is because you have the git submodules in your changes. That needs to be fixed before this can be merged. 

I believe that running `make git-submodule` and then adding a commit should fix this issue.



> Debian avr-gcc 5.4.0

If you can update to 8.x, you'll get a lot more savings in firmware size.

**Comment by josjoha** [comment]
Thanks, I'll look into 8.x later. I did try the sub-module thing, but the commit said there was nothing to commit.
```

/home/joshb/hardware/keyboard/minivan/git/qmk_firmware
josdebian% make git-submodule
QMK Firmware 0.9.58
Synchronizing submodule url for 'lib/chibios'
Synchronizing submodule url for 'lib/chibios-contrib'
Synchronizing submodule url for 'lib/googletest'
Synchronizing submodule url for 'lib/lufa'
Synchronizing submodule url for 'lib/printf'
Synchronizing submodule url for 'lib/ugfx'
Synchronizing submodule url for 'lib/vusb'
josdebian% git commit
On branch minivan_dvorak
Your branch is up to date with 'origin/minivan_dvorak'.

nothing to commit, working tree clean
```
I already had done the “make sub-module” thing before committing/pushing the above. Do I need to mark something as to be committed ?

**Comment by noroadsleft** [comment]
> It looks like you may need to run `make git-submodule` and add a new commit to fix the lib/ errors.
> 
> Otherwise, noroadsleft ?

Not sure how I missed this mention, and I deeply apologize for that.

I've been looking into this for the last few hours, and my advice is this:

From your `qmk_firmware` directory, run:

```shell
cd lib/chibios/
git checkout 313416b8fda90d9973a749a0a35970956852c286
cd ../../
cd lib/chibios-contrib/
git checkout e3a3a24047717af33b098266e11c2e4e5102206a
cd ../../
cd lib/ugfx/
git checkout 40b48f470addad6a4fb1177de1a69a181158739b
cd ../../
git add -- lib/chibios/ lib/chibios-contrib/ lib/ugfx/
git commit -m "fix modified submodules"
git push
```

A bit involved, but I have yet to figure out a simple, clean way to tell Git to check what the upstream repository (specifically, `qmk/qmk_firmware`) has for its submodules, and instruct the fork to use that.

**Comment by josjoha** [comment]
Thanks noroadsleft ! No need to apologize. I was working on it anyway. Oh it looks like the build failed. In any case I still want to add some stuff, shouldn't take too long now. When I'm done I figure I could create a new fork, paste in my keymap there, commit/push. I'll follow the advice of the QMK crew.

**Comment by josjoha** [comment]
It says here the build is too large, but it isn't too large here. It used to be ± 350-410 bytes left over. When I rebased to the current QMK, it is still 166 bytes free.

**Comment by noroadsleft** [comment]
Can confirm:

```shell
Checking file size of thevankeyboards_minivan_josjoha.hex                                           [OK]
 * The firmware size is fine - 27224/28672 (94%, 1448 bytes free)
```

The CI environment run compiles under avr-gcc 5.4.0, which is quite old and doesn't compile nearly as small. My environment is version 8.3.0; as you can see it's considerably smaller.

**Review by noroadsleft — COMMENTED** [review]
Submodules are fixed. :+1: Not diving into the keymap code (it's way over my head), but it compiles.

**Comment by josjoha** [comment]
Although I still fixed a coding mistake last week (speed precision), it seems stable. Some weeks ago I have gone through everything again, so it may not be needed to recheck everything. Once all done, some beta testing by someone else could be useful.

To do: add Workman, Qwerz, Azerty, Qwerty with arrows on top layouts, adding another layout overview, add compiling as single layout (relatively simple things at this point). I assume it will eventually have to compile correctly on git for it to be merged. Here it compiles with 5.4.0:
```
> uname -a
Linux josdebian 4.19.0-10-amd64 #1 SMP Debian 4.19.132-1 (2020-07-24) x86_64 GNU/Linux

> avr-gcc --version
avr-gcc (GCC) 5.4.0
(…)

> make thevankeyboards/minivan:josjoha
(…)
 * The firmware size is approaching the maximum - 28506/28672 (99%, 166 bytes free)
```

**Comment by noroadsleft** [comment]
> I assume it will eventually have to compile correctly on git for it to be merged.

Not true, actually. If we can confirm that it at least compiles locally, it can be merged. (We're aware the CI compiles larger.)

**Comment by josjoha** [comment]
> > I assume it will eventually have to compile correctly on git for it to be merged.
> 
> Not true, actually. If we can confirm that it at least compiles locally, it can be merged. (We're aware the CI compiles larger.)

Okidoki then I won't worry about it. I'll keep using 5.4.0 which compiles larger, gives it all a bit more playing room. I don't think i will need any more bytes anyway. I'll add a couple of more layout options, then we can finally merge it I hope.

**Comment by josjoha** [comment]
This pull request is _not_ stale. I worked on the code yesterday, improving the configurability of the Numbers Pad dual layout compile option. There is a delay in activity, due to external reasons, which should not be a significant problem for completing this project.

My plan is to add three more dual layout options: Qwerty with arrows on top (which could work well with regular Qwerty), Azerty (French) and Qwertz (German). The graphical documentation is usually most of the work for adding dual layouts.

**Comment by josjoha** [comment]
Looks like it is done. It is always possible to go over everything one more time, fix language mistakes and so on, but it seems to be basically ready for a review, and merging, if you like.

**Comment by josjoha** [comment]
haha genius, the automated check suddenly passes. Wew !

**Review by drashna — APPROVED** [review]

**Comment by josjoha** [comment]
@tzarc  seems to have left a request for changes, which is lingering. This is old and was dealt with back then.

**Comment by josjoha** [comment]
Found an unexpected problem, can't see a solution right now, although it is minor. It seems to be a sort of layer stuck problem. When compiled with Qwerty Base Arrow on _DEF_BASE, the key which holds to _RAR in most layouts, does not return to Base. It seems that the transparency on _RAR on that key, does not look to _ALT_BASE (layer 1), but rather _DEF_BASE (layer 0), where in this case that key is an arrow. I tried to resolve the issue with default_layer_set(layer_mask), but no luck yet.

It is a small wrinkle, only happens in one particular compile setting (Base Arrow on _DEF_BASE). It effectively changes a hold-layer to a toggle-layer, which isn't much of an issue (in this case). However I do want to fix it (obviously). (I've asked on #programming).

**Comment by josjoha** [comment]
Oops found a graphics documentation mishap for Qwerty Base Arrow, will fix tomorrow. Swwy.

**Comment by josjoha** [comment]
There are no changes requested. This seems to be an error in the git system or is not correctly identified as being resolved. As far as I can see the material can be merged.

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
> There are no changes requested. This seems to be an error in the git system or is not correctly identified as being resolved. As far as I can see the material can be merged.

This is a GitHub thing – when a maintainer requests changes, they remain as having requested changes until they have either approved the pull request, or had their review dismissed. This is the case even when the requested changes have been applied.

Looks good to me; merging!

**Comment by josjoha** [comment]
Hey man, thank you ! !!!

**Comment by josjoha** [comment]
Oh no I forgot to uncomment one thing in user_config.h grrrrr
It isn't critical or anything, but a user might want the default to compile from user_config.h not user_config_compact.h

**Comment by noroadsleft** [comment]
We accept follow-up PRs :)

**Pull request merged**
